theano/tensor/tests/test_opt.py

```diff
@@ -2995,6 +2995,33 @@ def test_incsubtensor_allocs1(self):
         assert numpy.all([not isinstance(n.op, tensor.IncSubtensor)
                           for n in f.maker.fgraph.toposort()])
 
+    def test_incsubtensor_x_zeros(self):
+        x = tensor.constant(numpy.asarray(numpy.zeros((4, 4)),
+                                          dtype=config.floatX))
+        y = tensor.matrix()
+        z = tensor.inc_subtensor(x[:4], y)
+        f = theano.function([y], z)
+        inc_nodes = [n for n in f.maker.fgraph.toposort()
+                     if isinstance(n.op, tensor.IncSubtensor)]
+
+        assert(len(inc_nodes) == 1)
+        node_is_set_instead_of_inc = inc_nodes[0].op.set_instead_of_inc
+        mode = theano.config.mode
+        assert((mode != "FAST_COMPILE" and node_is_set_instead_of_inc) or
+               (mode == "FAST_COMPILE" and not node_is_set_instead_of_inc))
+
+        # also check the flag doesn't get set if first input is not zeros:
+        not_all_zeros = numpy.zeros((4, 4))
+        not_all_zeros[2, 3] = 0.001
+        x = tensor.constant(numpy.asarray(not_all_zeros, dtype=config.floatX))
+        y = tensor.matrix()
+        z = tensor.inc_subtensor(x[:4], y)
+        f = theano.function([y], z)
+        inc_nodes = [n for n in f.maker.fgraph.toposort()
+                     if isinstance(n.op, tensor.IncSubtensor)]
+        assert(len(inc_nodes) == 1)
+        assert(inc_nodes[0].op.set_instead_of_inc is False)
```

Comment:
Execute both test function and check the result. Just in case, if we compile executing them won't add much tests time.